imap_processing/ultra/l0/decom_ultra.py

```diff
@@ -1,211 +1,139 @@
+from imap_processing.ultra.l0.decom_tools import (
+    decompress_binary,
+    decompress_image,
+    read_image_raw_events_binary,
+)
+from imap_processing.ultra.l0.ultra_utils import ParserHelper, UltraParams
+from imap_processing.utils import group_by_apid, sort_by_time
 
 logging.basicConfig(level=logging.INFO)
 
 
-class PacketProperties(NamedTuple):
-    """Class that represents properties of the ULTRA packet type."""
-
-    apid: int
-    width: int
-    block: int
-    len_array: int
-
-
-class UltraParams(Enum):
-    """Enumerated packet properties for ULTRA."""
-
-    ULTRA_AUX = PacketProperties(apid=880, width=None, block=None, len_array=None)
-    ULTRA_IMG_RATES = PacketProperties(apid=881, width=5, block=16, len_array=48)
-
-
-def read_n_bits(binary: str, n: int, current_position: int):
-    """Extract the specified number of bits from a binary string.
-
-    Starting from the current position, it reads n bits. This is used twice.
-    The first time it reads the first 5 bits to determine the width.
-    The second time it uses the width to determine the value of the bitstring.
-
-    Parameters
-    ----------
-    binary : str
-        The string of binary data from which bits will be read.
-        This is a string of 0's and 1's.
-    n : int
-        Number of bits to read from the binary string.
-    current_position : int
-        The starting position in the binary string from which bits will be read.
-
-    Returns
-    -------
-    value : int
-        The integer representation of the read bits or None if the end of the
-        string is reached before reading 'n' bits.
-    current_position + n
-        - The updated position in the binary string after reading the bits.
-    """
-    # Ensure we don't read past the end
-    if current_position + n > len(binary):
-        raise IndexError(
-            f"Attempted to read past the end of binary string. "
-            f"Current position: {current_position}, "
-            f"Requested bits: {n}, String length: {len(binary)}"
-        )
-
-    value = int(binary[current_position : current_position + n], 2)
-    return value, current_position + n
-
-
-def log_decompression(value: int) -> int:
+def append_params(
+    decom_data: dict, packet, decompressed_data=None, decompressed_key=None
+):
     """
-    Perform logarithmic decompression on a 16-bit integer.
+    Append parsed items to a dictionary, including decompressed data if available.
 
     Parameters
     ----------
-    value : int
-        A 16-bit integer comprised of a 4-bit exponent followed by a 12-bit mantissa.
-
-    Returns
-    -------
-    int
-        The decompressed integer value.
-
-    Note: Equations from Section 1.2.1.1 Data Compression and Decompression Algorithms
-    in Ultra_algorithm_doc_rev2.pdf.
+    decom_data : dict
+        Dictionary to which the data is appended.
+    packet : space_packet_parser.parser.Packet
+        Individual packet.
+    decompressed_data : list
```

Comment:
You are appending this entire list to the key below, otherwise, you're just appending a single item from the packet derived value. Do you need to index into this decompressed data to pick off the single value you want associated with a specific key?

I think I might just be confused on the flow through here and would have expected a function call on the item data below rather than passing it in here. But, I'm just commenting here in case that brings up any ideas for you, this isn't anything blocking/major.

I expected something like:
```python
if key in decompressed_keys:
    # We have a field that needs to be decompressed
    decom_data.append(call_decompression_function(item.raw_value))
```